### PR TITLE
Fix/3.0.0 fixes

### PR DESCRIPTION
**Pull request by InfiniteLee**
This is a followup to #79 and #82.

My original goal with #79 was to enable a CANNON body that was `type: dynamic` and `mass: 0`, effectively making it "kinematic" (as defined in other game engines like Unity). That is to say, it imparts force on other `dynamic` bodies, collides with both `dynamic` and `static` bodies, and can be moved around by directly setting position. (Note: CANNON.js has a defined `kinematic` body type, but it only allows movement via velocity, and doesn't collide with other `static` or `kinematic` bodies. So I will refer to a `dynamic` body with `mass: 0` as "kinematic" in quotes). 
My other goal with #79 was to allow the mass value to change, effectively allowing the body to switch from `dynamic` and "kinematic" on the fly. 

To support this post 3.0.0, a few changes are needed.

- type needs to be explicitly set when creating the CANNON body. Otherwise it will infer type from the mass, and as such will always set type to `dynamic` if `mass != 0` and `static` if `mass == 0`.
- Don't update the mass in `update` when explicitly switching between `dynamic` and `static`
- use the body's mass when determining whether to use `syncToPhysics` or `syncFromPhysics` in `beforeStep` and `step` instead of the component's `data.type`.

Lastly, there are a few bug fixes: 

- Since the quaternion component is gone now, rotation and position are handled differently in 3.0.0. However, `quaternion` can't be updated directly on the object3d as that won't update the rotation component on the Aframe side. As such, go back to using setAttribute, but do some math for converting the quaternion to euler angles in degrees. (Unfortunate because this effectively gets undone under the hood to ultimately set the object3d's quaternion, but that's Aframe.)
- The fix for #82

### CODE REVIEW

File: src/components/body/body.js [hunk 1]
```diff
@@ -127,16 +128,6 @@ var Body = {
...
 
-    if (data.type !== prevData.type) {
-      if (data.type === 'dynamic' && data.mass === 0) {
-        this.el.setAttribute(this.name, {mass: 5});
-      }
-      if (data.type === 'static' && data.mass !== 0) {
-        this.el.setAttribute(this.name, {mass: 0});
-      }
-      return;
-    }
```
Comment (donmccurdy) [inline src/components/body/body.js]: i expect we need to do something if `data.type` changes, no? Not sure if `body.type` can actually be changed after the fact, if not we should log a warning...

File: src/components/body/body.js [hunk 2]
```diff
@@ -276,18 +268,22 @@ var Body = {
...
 
       if (parentEl.isScene) {
-        el.object3D.quaternion.copy(body.quaternion);
-        el.object3D.position.copy(body.position);
+        q1.copy(body.quaternion);
+        e.setFromQuaternion(q1, 'YXZ');
+        el.setAttribute('rotation', { x: e.x * THREE.Math.RAD2DEG, 
+                                      y: e.y * THREE.Math.RAD2DEG, 
+                                      z: e.z * THREE.Math.RAD2DEG });
+        el.setAttribute('position', body.position);
```
Comment (donmccurdy) [inline src/components/body/body.js]: what version of A-Frame are you working with? I'd removed the `quaternion` component because A-Frame 0.8.0 will allow you to write to`object3D.*` safely. Using an euler will give gymbal lock issues (or i think that's the term for it). If we need this fixed pre-0.8.0 i may need to restore `quaternion`...

File: src/components/body/body.js [hunk 3]
```diff
@@ -276,18 +268,22 @@ var Body = {
...
       } else {
         // TODO - Nested rotation doesn't seem to be working as expected.
         q1.copy(body.quaternion);
         parentEl.object3D.getWorldQuaternion(q2);
         q1.multiply(q2.inverse());
-        el.object3D.quaternion.copy(q1);
+        el.setAttribute('rotation', {x: q1.x, y: q1.y, z: q1.z, w: q1.w});
```
Comment (donmccurdy) [inline src/components/body/body.js]: see above.

File: src/components/body/body.js (hunk 2 above)
Comment (InfiniteLee) [reply src/components/body/body.js]: I'm using a specific hash of aframe > 0.7.0 but should be able to be caught up to master soon. That said, it's pretty broken for <= 0.7.0, but its up to you if you want >= 3.0.0 to be compatible with <= 0.7.0.
Also, I'm not sure the way I'm using an euler will actually cause gimbal lock or not- I suspect not though because I'm not actually rotating the euler, just converting the quaternion to a euler.

File: src/components/body/body.js (hunk 1 above)
Comment (InfiniteLee) [reply src/components/body/body.js]: I don't think you can actually change the type. The only way to effectively switch between dynamic and static is by starting with dynamic and changing the mass from non-zero to zero. So you are probably right, a warning should be logged.

**Comment by InfiniteLee** [comment]
The problem with the Cannon `kinematic` bodies is less so not being able to move them around, but more so that they don't collide with other `static` and `kinematic` bodies.

Let me rephrase what I said earlier, because I think by using the term "kinematic" I confused things a bit.
What I really want to be able to do is switch between a `dynamic` and `static` body, on the fly. The problem is you can't do that in Cannon . In other engines, like Ammo, you can, with a few small tricks. What I did realize, however, was that the only thing Cannon 's `body` actually uses to determine if a body should behave dynamically or statically is if the mass is zero or not. As such, you can effectively switch between `dynamic` and `static` by changing the mass. 

The part about being a "kinematic" body is that what Cannon calls `static` is actually what is considered `kinematic` in other physics engines I've seen (e.g. Unity's physics). So sorry to confuse things there.
So, that said, is this exposing an edge case in Cannon? Perhaps. But its also the way schteppe has said to do it: https://github.com/schteppe/cannon.js/issues/317#issuecomment-262710586 so it seems reasonable to be able to support it, since we're not likely to see an improved API in Cannon anytime soon. 

As far as using a different engine, I originally wanted to fork this project and try to recreate it with Ammo, but decided it would be a lot less work to start with what was here already. That said, I fully support (and would happily contribute to) any such efforts to switch to a different physics engine.
